go/vt/tlstest/tlstest.go

```diff
@@ -19,207 +19,362 @@ limitations under the License.
+	keyData, err := x509.MarshalPKCS8PrivateKey(key)
+	if err != nil {
+		return err
+	}
+	out := &bytes.Buffer{}
+	err = pem.Encode(out, &pem.Block{Type: "PRIVATE KEY", Bytes: keyData})
+	if err != nil {
+		return err
+	}
+	return os.WriteFile(keyPath, out.Bytes(), permissions)
+}
+
+// pubkey is an interface to get a public key from a private key
+// The Go specification for a private key defines that this always
+// exists, although there's no interface for it since it would break
+// backwards compatibility. See https://pkg.go.dev/crypto#PrivateKey
+type pubKey interface {
+	Public() crypto.PublicKey
+}
+
+func publicKey(priv crypto.PrivateKey) crypto.PublicKey {
+	return priv.(pubKey).Public()
+}
+
+func signCert(parent *x509.Certificate, parentPriv crypto.PrivateKey, certPub crypto.PublicKey, commonName string, serial int64, ca bool) (*x509.Certificate, error) {
+	keyUsage := x509.KeyUsageDigitalSignature
+	var extKeyUsage []x509.ExtKeyUsage
+	var dnsNames []string
+	var ipAddresses []net.IP
+
+	if ca {
+		keyUsage = keyUsage | x509.KeyUsageCRLSign | x509.KeyUsageCertSign
+	} else {
+		extKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth}
+		dnsNames = []string{"localhost", commonName}
+		ipAddresses = []net.IP{net.ParseIP("127.0.0.1"), net.ParseIP("::1")}
+	}
+
+	template := x509.Certificate{
+		SerialNumber: big.NewInt(serial),
+		Subject: pkix.Name{
+			CommonName: commonName,
+		},
+		NotBefore:             time.Now().Add(-30 * time.Second),
+		NotAfter:              time.Now().Add(24 * time.Hour),
+		KeyUsage:              keyUsage,
+		ExtKeyUsage:           extKeyUsage,
+		BasicConstraintsValid: true,
+		IsCA:                  ca,
+		DNSNames:              dnsNames,
+		IPAddresses:           ipAddresses,
 	}
 
-	config := path.Join(root, parent+"-ca.config")
-	if _, err := os.Stat(config); os.IsNotExist(err) {
-		if err := os.WriteFile(config, []byte(fmt.Sprintf(caConfigTemplate, databasePath)), os.ModePerm); err != nil {
-			log.Fatalf("cannot write file %v: %v", config, err)
-		}
+	// No parent defined means we create a self signed one.
+	if parent == nil {
+		parent = &template
 	}
+
+	certificate, err := x509.CreateCertificate(rand.Reader, &template, parent, certPub, parentPriv)
+	if err != nil {
+		return nil, err
+	}
+	return x509.ParseCertificate(certificate)
 }
 
 // CreateCA creates the toplevel 'ca' certificate and key, and places it
 // in the provided directory. Temporary files are also created in that
 // directory.
 func CreateCA(root string) {
 	log.Infof("Creating test root CA in %v", root)
-	key := path.Join(root, "ca-key.pem")
-	cert := path.Join(root, "ca-cert.pem")
-	config := path.Join(root, "ca-ca.config")
-	createKeyDBAndCAConfig(root, "ca")
-
-	openssl("genrsa", "-out", key)
-
-	openssl("req",
-		"-new",
-		"-x509",
-		"-days", "3600",
-		"-nodes",
-		"-batch",
-		"-sha256",
-		"-config", config,
-		"-key", key,
-		"-out", cert)
+	keyPath := path.Join(root, "ca-key.pem")
```

Comment:
For files, you specifically want `filepath.Join` as that will also work on Windows. See the package doc https://pkg.go.dev/path/filepath.

Reply:
@mdlayher I haven't updated any of this and followed the pattern that's already there for now. I doubt Vitess on Windows is even supported at all.